trace_decoder/src/compact/compact_to_partial_trie.rs

```diff
@@ -8,183 +11,261 @@ use mpt_trie::{
+        curr_key: Nibbles,
+        ext_node_key: &Nibbles,
+        ext_child: &NodeEntry,
+    ) -> CompactParsingResult<()> {
+        let new_k = curr_key.merge_nibbles(ext_node_key);
+        create_partial_trie_from_compact_node_rec(new_k, ext_child, self)?;
 
-    match curr_node {
-        NodeEntry::Branch(n) => process_branch(curr_key, n, output),
-        NodeEntry::Code(c_bytes) => process_code(c_bytes.clone(), output),
-        NodeEntry::Empty => process_empty(),
-        NodeEntry::Hash(h) => process_hash(curr_key, *h, &mut output.trie),
-        NodeEntry::Leaf(k, v) => process_leaf(curr_key, k, v, output),
-        NodeEntry::Extension(k, c) => process_extension(curr_key, k, c, output),
+        Ok(())
     }
+
+    fn get_trie(&mut self) -> &mut HashedPartialTrie;
 }
 
-fn process_branch(
-    curr_key: Nibbles,
-    branch: &[Option<Box<NodeEntry>>],
-    output: &mut CompactToPartialTrieExtractionOutput,
-) -> CompactParsingResult<()> {
-    for (i, slot) in branch.iter().enumerate().take(16) {
-        if let Some(child) = slot {
-            // TODO: Seriously update `mpt_trie` to have a better API...
-            let mut new_k = curr_key;
-            new_k.push_nibble_back(i as Nibble);
-            create_partial_trie_from_compact_node_rec(new_k, child, output)?;
+#[derive(Debug)]
+pub(super) enum UnexpectedCompactNodeType {
+    AccountLeaf,
+    Code,
+}
+
+impl Display for UnexpectedCompactNodeType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            UnexpectedCompactNodeType::AccountLeaf => write!(f, "AccountLeaf"),
+            UnexpectedCompactNodeType::Code => write!(f, "Code"),
         }
     }
+}
 
-    Ok(())
+#[derive(Debug, Default)]
+pub(super) struct StateTrieExtractionOutput {
+    pub(super) trie: HashedPartialTrie,
+    pub(super) code: HashMap<CodeHash, Vec<u8>>,
+    pub(super) storage_tries: HashMap<HashedAccountAddr, HashedPartialTrie>,
 }
 
-fn process_code(
-    c_bytes: Vec<u8>,
-    output: &mut CompactToPartialTrieExtractionOutput,
-) -> CompactParsingResult<()> {
-    let c_hash = hash(&c_bytes);
-    output.code.insert(c_hash, c_bytes);
+impl CompactToPartialTrieExtractionOutput for StateTrieExtractionOutput {
+    fn process_code(&mut self, c_bytes: Vec<u8>) -> CompactParsingResult<()> {
+        let c_hash = hash(&c_bytes);
+        self.code.insert(c_hash, c_bytes);
 
-    Ok(())
-}
+        Ok(())
+    }
 
-fn process_empty() -> CompactParsingResult<()> {
-    // Nothing to do.
-    Ok(())
+    fn process_leaf(
+        &mut self,
+        curr_key: Nibbles,
+        leaf_key: &Nibbles,
+        leaf_node_data: &LeafNodeData,
+    ) -> CompactParsingResult<()> {
+        process_leaf_common(
+            &mut self.trie,
+            curr_key,
+            leaf_key,
+            leaf_node_data,
+            |acc_data, full_k| {
+                Ok(
+                    convert_account_node_data_to_rlp_bytes_and_add_any_code_to_lookup(
```

Comment:
Gosh this is a long method name

Reply:
You haven't seen any of my 100+ char method names yet! :sweat_smile: I'll see if I can shorten it.

Reply:
Now `process_account_node`.